### ISSUE TITLE
Unnecessary preemption during DaemonSet rolling update

### ISSUE THREAD

**Issue by johngmyers** [kubernetes/kubernetes#102382]
#### What happened:

On a cluster with a 3-node control plane, had worker nodes each with a number of high priority DaemonSet pods and a single normal priority StatefulSet pod. The nodes had 220m CPU unallocated.

Updated the spec of a high priority DaemonSet with a 700m CPU request (no change to requests), triggering a rolling update of that DaemonSet.

On each successive node, first the old DaemonSet pod was terminated, then the StatefulSet pod was terminated. Then the new DaemeonSet pod was scheduled, then a replacement StatefulSet pod was scheduled. API server audit logs showed that the StatefulSet pods were terminated by kube-scheduler.

#### What you expected to happen:

I expected the StatefulSet pods to not be terminated.

#### How to reproduce it (as minimally and precisely as possible):

I have not yet reproduced it.

#### Anything else we need to know?:

I suspect this was caused by propagation delay of the removal event for the old DaemonSet pod to kube-scheduler, relative to the event for the creation of the replacement DaemonSet pod.

#### Environment:
- Kubernetes version (use `kubectl version`): 1.17.17
- Cloud provider or hardware configuration: AWS
- OS (e.g: `cat /etc/os-release`): Ubuntu 20.04.2
- Kernel (e.g. `uname -a`): 5.4.0-1048-aws #50-Ubuntu SMP
- Install tools: kops
- Network plugin and version (if this is a network-related bug):
- Others:

/sig scheduling

**Comment by ahg-g** [comment]
> I suspect this was caused by propagation delay of the removal event for the old DaemonSet pod to kube-scheduler, relative to the event for the creation of the replacement DaemonSet pod.

That is possible.

**Comment by johngmyers** [comment]
> That is possible.

Would the resulting unnecessary preemption, despite PDB, be correctly classified as behavior contrary to the documented expectations of the system?

As for addressing this, perhaps terminating pods should not contribute toward preemption? (assuming they do; I have not read that code)

I will spend more time next week seeing if I can reproduce.

**Comment by ahg-g** [comment]
> Would the resulting unnecessary preemption, despite PDB, be correctly classified as behavior contrary to the documented expectations of the system?

The scheduler respect PDB on best effort basis only (see https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/#poddisruptionbudget-is-supported-but-not-guaranteed). Which documentation are you referring to?

**Comment by johngmyers** [comment]
https://kubernetes.io/docs/tasks/manage-daemon/update-daemon-set/#daemonset-update-strategy

> At most one pod of the DaemonSet will be running on each node during the whole update process.

Kubernetes is behaving as if both pods of the DaemonSet were simultaneously running on the node.

**Comment by ahg-g** [comment]
> https://kubernetes.io/docs/tasks/manage-daemon/update-daemon-set/#daemonset-update-strategy
> 
> > At most one pod of the DaemonSet will be running on each node during the whole update process.
> 
> Kubernetes is behaving as if both pods of the DaemonSet were simultaneously running on the node.

Not really, it is still true that only one pod is running **on the node**.

**Comment by johngmyers** [comment]
If only one pod was running on the node then the scheduler had no justification for preempting the StatefulSet.

The Kubernetes system as a whole needs to maintain the invariants that the Kubernetes system documents, despite such implementation concerns as propagation delay.

**Comment by ahg-g** [comment]
Kubernetes is eventually consistent. The controllers driver the system state to eventually match the user declared one.

The invariant "At most one pod of the DaemonSet will be running on each node during the whole update process." isn't being violated.

**Comment by johngmyers** [comment]
kube-scheduler is taking action based on a view of the state of the world that never existed. As such, it is performing a preemption contrary to the documented bases.

The usual way of dealing with this sort of issue is with conditional requests. For example, kube-scheduler would need to be able to tell apiserver "delete this pod, but only if all these other pods still exist".

**Comment by ahg-g** [comment]
> kube-scheduler is taking action based on a view of the state of the world that never existed.

The scheduler is taking action based on the stream of events it is receiving from the api-server.

> As such, it is performing a preemption contrary to the documented bases.

I still fail to understand which "documentation" the scheduler is not adhering to?

> The usual way of dealing with this sort of issue is with conditional requests. For example, kube-scheduler would need to be able to tell apiserver "delete this pod, but only if all these other pods still exist".

Feel free to make a full proposal. But my guess is that the complexity that comes with such approach is not worth it.

One thing you may want to look at is `PreemptionPolicy: Never` as a way to prevent preemption during daemonset upgrades.

**Comment by johngmyers** [comment]
`PreemptionPolicy: Never` would have to be put on the `system-node-critical` PriortityClass (and all other PriorityClasses used by daemonsets).

The scheduler is not adhering to the documentation stating that "At most one pod of the DaemonSet will be running on each node during the whole update process." It is evicting workloads due to the incorrect belief that there are two pods of the DaemonSet running on the node.

/remove-lifecycle rotten
